### PR TITLE
Add risk group

### PR DESCRIPTION
**Pull request by Fuhan-Yang**
- Added RiskBin
- Added examples/sir_age_risk_stratified.py

In the example, three age groups (young, adult, elderly) and two risk group (high, low) are combined to form an age-risk combinations. The initial proportion of S and I, force of infection were created based on this. 

Given age proportion to be: ["young": 0.7, "adult": 0.2,"elderly": 0.1]; high-low risk proportion for each age group: ["young": [0.1,0.9], "adult": [0.6,0.4], "elderly": [0.8,0.2]], and contact matrix is: 

image

The dynamics will look like: 
image

### CODE REVIEW

File: examples/sir_age_risk_stratified.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+class SIRInitializer(Initializer):
+    """Initializer for SIR model, setting initial conditions for compartments."""
+
+    model_config = ConfigDict(arbitrary_types_allowed=True, extra="allow")
+
+    description: str = "An SIR initializer"
+    initialize_date: date = date(2022, 2, 11)  # random date
+    population_size: int = 1000
+
+    age_demographics: jnp.ndarray = Field(...)
```
Comment (kokbent) [inline examples/sir_age_risk_stratified.py:34]: Not wrong per se... but this feels very uncommon and unrestricted like i just learned about this Field thing from pydantic from here... unless there's a specific reason for it, i suggest we use the more familiar class pattern using `__init__`:
```
    def __init__(self, age_demographics, risk_prop):
        super().__init__(
            description=...,
            initialize_date=...,
            population_size=...,
        )
        self.age_demographics = age_demographics
...
```
See other examples for more of using the initialization style. I think one issue with using the Field is you can initialize a class without having any argument in those attributes? and it would be initialize as `None` and can get messy? Not tooooo sure what are the implications.

File: examples/sir_age_risk_stratified.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        """Get initial compartment values for an SIR model stratified by age."""
+        assert s0_prop + i0_prop == 1.0, (
+            "s0_prop and i0_prop must sum to 1.0, "
+            f"got {s0_prop} and {i0_prop}."
+        )
+
+        age_risk_prop = (
+            self.age_demographics[:, None] * self.risk_prop
+        )  # age by row, risk by column
+        num_susceptibles = self.population_size * jnp.array([s0_prop])
```
Comment (kokbent) [inline examples/sir_age_risk_stratified.py]: `num_susceptibles` and `num_infectious` are scalars right? Probably don't need to cast them as `jnp.array`

File: examples/sir_age_risk_stratified.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+            [
+                [[0.09, 0.21], [0.06, 0.14], [0.27, 0.23]],
+                [[0.33, 0.11], [0.05, 0.07], [0.21, 0.23]],
+            ],
+            [
+                [[0.16, 0.24], [0.08, 0.06], [0.22, 0.24]],
+                [[0.19, 0.01], [0.28, 0.17], [0.14, 0.21]],
+            ],
+        ]
+    )
```
Comment (kokbent) [inline examples/sir_age_risk_stratified.py:115]: This is a pretty complicated "matrix"... and it's hard to see if it's "appropriate" (e.g., that they are properly normalized, like in 2d matrix, we divide the matrix with the max of the eigenvalues, what could we do here to ensure that here?) and how one could construct it...

Maybe we'll need to think through examples of how you construct this 3x2x3x2 array. Lets say age contact is [[0.7, 0.2, 0.1], [0.2, 0.7, 0.1], [0.1, 0.1, 0.8]], and assume equal contact of [[0.5, 0.5], [0.5, 0.5]] for risk contacts... how do we construct an age_risk_contact that is normalized?

**Review by kokbent — COMMENTED** [review]
I have not finished review through the ODE part yet, will work on it when i have time

File: examples/sir_age_risk_stratified.py (hunk 3 above)
Comment (ElishaBayode) [reply examples/sir_age_risk_stratified.py:115]: I agree with Ben. One way to deal with this is to avoid writing the full tensor directly (or manually) and instead construct it as the Kronecker product of two lower-dimensional contact matrices (one for age based contact and the other for risk based contact), so you have a 3x3 age-age contact matrix and a 2x2 risk-risk contact matrix. If each of these matrices is row-stochastic (each row sums to 1, you can confirm this easily), then their Kronecker product defines a normalized age-risk contact operator (after reshaping back to 3×2×3×2), with max eigenvalue = 1.

File: src/dynode/config/bins.py [hunk 4]
```diff
@@ -87,3 +87,9 @@ class WaneBin(Bin):
         description="""Proportion of immune protection retained by populations within
           this bin, between [0-1]. May be subsequently modified by strain immune escape.""",
     )
+
+
+class RiskBin(Bin):
```
Comment (kokbent) [inline src/dynode/config/bins.py]: Not a big fan of having RiskBin in the Dynode level (it doesn't have special notion like wanebin with its protection specification, or AgeBin with age demarcation)... See corresponding comment in`sir_age_risk_stratified`

File: examples/sir_age_risk_stratified.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    age_dimension = Dimension(
+        name="age",
+        bins=[
+            AgeBin(0, 17, name="young"),
+            AgeBin(18, 64, name="adult"),
+            AgeBin(65, 99, name="elderly"),
+        ],
+    )
+    risk_dimension = Dimension(
+        name="risk", bins=[RiskBin(name="high"), RiskBin(name="low")]
```
Comment (kokbent) [inline examples/sir_age_risk_stratified.py]: The bins here are only for naming, so perhaps we don't need specific class `RiskBin`... can just do...
`name="risk", bins=[Bin(name="high"), Bin(name="low")]`

File: examples/sir_age_risk_stratified.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+@jax.jit
+def sir_ode(
+    t: float, state: CompartmentState, p: SIR_ODEParams
+) -> CompartmentGradients:
+    """A simple SIR ODE model with no time-varying components."""
+    s, i, r = state
+    pop_size = s + i + r
+    force_of_infection = p.beta * jnp.einsum(
+        "ijkl,mn -> ij", (p.contact_matrix) / pop_size, i
```
Comment (kokbent) [inline examples/sir_age_risk_stratified.py]: This (`ijkl,mn -> ij`) doesn't look right to me... this is saying take the first array, reduce the dimension to two (keeping ij dimension and collapse kl matrix). As a result, you get only a calculation that doesn't involve `i` compartment at all...

Comment (kokbent) [reply examples/sir_age_risk_stratified.py]: I suspect what you wanted is closer to `ijkl,kl->ij`... but check the math, set `i` to `[[0, 1], [0, 0], [0, 0]]` to see if it yields what you want.

Comment (ElishaBayode) [inline examples/sir_age_risk_stratified.py]: There is a mismatch in the Einsteinsum specification, the indices don’t align e.g., kl should correspond to the infectious indices mn. Also, the FOI is not written in the standard form: the contact matrix should be contracted with the infectious fraction (infectors), so the normalization should divide by the infectious group population, not the susceptible group population.

**Review by ElishaBayode — COMMENTED** [review]
I will go over it again when I have more time

**Comment by kokbent** [comment]
On another note, I think it's best to go back to just age x age contact matrix first, note that in flu model we also have age x vaccination dimension similar to the concept of age x risk here. In that we assume equal mixing in vaccination status. You could check that model out and make sure the differentials part is correctly specified before venture into creating a 4d contact matrix. And yeah the kronecker product idea by Elisha seems to be right...

File: examples/sir_age_risk_stratified.py (hunk 1 above)
Comment (Fuhan-Yang) [reply examples/sir_age_risk_stratified.py:34]: Using the code you showed here will return an error for me when creating an instance of SIRInitializer(): `"SIRInitializer" object has no field "age_demographics"`. Maybe this is because `Initializer` is a BaseModel and requires the members to be defined by `Field()`. Also, when instantiating `SIRInitializer`, it will throw validation errors if either `age_demographics` or `risk_prop` is undefined or defined as other types than `jnp.array`, which shows the strength of validation of pydantic!

Can you show me what have you learned about Field that indicates this is not a common thing?

File: examples/sir_age_risk_stratified.py (hunk 2 above)
Comment (Fuhan-Yang) [reply examples/sir_age_risk_stratified.py]: Yes they are scalars. Sounds good to me

File: examples/sir_age_risk_stratified.py (hunk 3 above)
Comment (Fuhan-Yang) [reply examples/sir_age_risk_stratified.py:115]: Thanks for pointing that out. I've made changes to calculate the 3x2x3x2 array based on the 3x3 age-based contact matrix and 2x2 risk-based contact matrix. For each 3x2 block matrix, it means the contact rate across one-bin age and one-bin risk. For example, the age bins are [young, adult, elderly], and the risk bins are [high,low]. For the block matrix at [0,0,:,:], it is the contact rate for the young and high population to other populations. 

I still make unequal risk contact...this is easier for me to check if the calculation is correct. Like the results will kinda hide from me with the same risk contact rate. 

I do have a newbie question...why did we still normalize the contact matrices when they are row-stochastic? Like it already represents the probability of contacting other groups for this certain group, removing the effect of denominator (number of population reported, surveyed...etc), can you show me some examples of why, and how dividing the dominant eigenvalue can achieve that? Thanks!!

File: examples/sir_age_risk_stratified.py (hunk 6 above)
Comment (Fuhan-Yang) [reply examples/sir_age_risk_stratified.py]: You are right. On the second look of this, when correcting the contact matrix, it should be `ijkl, ij -> ij`, as we are summing up all the contact rate for $group_{i,j}$ and multiply with its corresponding $I_{i,j}$

Comment (Fuhan-Yang) [reply examples/sir_age_risk_stratified.py]: You are right. I corrected this to be: 
`force_of_infection = p.beta * jnp.einsum(
        "ijkl,ij -> ij", (p.contact_matrix) / pop_size, i
    )
`, so it's calculating: 

$$
FOI_{ij} = \beta\frac{ \sum_k\sum_l c_{ijkl} \times I_{ij}}{N}
$$

which is summing up all the contacts made by infected population $ij$ (say young and high risk) and multiplying the number of infections in that group, then the rate from S to I will be $S_{ij} \times I_{ij}$. Please see the updates!

**Comment by Fuhan-Yang** [comment]
> On another note, I think it's best to go back to just age x age contact matrix first, note that in flu model we also have age x vaccination dimension similar to the concept of age x risk here. In that we assume equal mixing in vaccination status. You could check that model out and make sure the differentials part is correctly specified before venture into creating a 4d contact matrix. And yeah the kronecker product idea by Elisha seems to be right...

That's a good suggestion! After I made changes per the comments, I run the model given the same age and risk proportion ([0.5,0.5]) and same contact matrix for age and risk (both are [[0.5,0.5],[0.5,0.5]]), I got the exactly same trajectories for S, I, R, as expected:

image

**Comment by Fuhan-Yang** [comment]
I also refactor the runtime tests to check the shape of several inputs to make sure the contact matrices and age and risk proportion can be correctly calculated. Note that the checks require that the bins from age or risk to be at least 2. Like multiple ages with 1 risk or 1 age with multiple risks won't work here (for the use of `jnp.einsum` ). Please let me know that these cases are necessary then I can make relevant change.

**Comment by Fuhan-Yang** [comment]
The updated example and test are ready to review. Added test to make sure the returned initial proportion and contact matrix by age and risk have the expected shape and values, given multiple bins and single bin.  

Here is how the 4-D contact matrix is derived from 2-D age-specific and risk-specific contact matrix. 
image

File: examples/sir_age_risk_stratified.py (hunk 3 above)
Comment (ElishaBayode) [reply examples/sir_age_risk_stratified.py:115]: Row-stochastic matrices need other conditions for them to automatically have spectral radius equal to 1 (which might not be met, like stationarity or being a transition matrix), so it's safe to normalize by the spectral radius to ensure that your r0 parameter correctly represents the basic reproduction number in the model. Not doing so may inadvertently amplify your r0

File: examples/sir_age_risk_stratified.py (hunk 1 above)
Comment (kokbent) [reply examples/sir_age_risk_stratified.py:34]: Mostly because I (and presumably others) don't come into using the package knowing the specifics of what `pydantic` is capable or not, and definitely not the use of `Field`. I tested the usage of `Field()` in a barebone example and it did not perform any validation but I've now checked in this specific case it indeed throws validation errors. This is probably due to the class itself is a pydantic class so the usage of `Field()` is well supported. I'm not against using it here, might just need brief comment on what it does since this is an example.

**Comment by kokbent** [comment]
@Fuhan-Yang Something is not quite right...

So I took the code and modified such that I can specify the initial state in more details (and I highly recommend you to implement that instead of the current one which doesn't allow you to specify which age-risk group has or has no infecteds). I made it such that 1% of Young-High and 1% of Adult-High started out with infected, all other groups have 0% infected.

The config is as follows, such that there are equal proportion of risk across all ages, there are mixing across Young and Adult but Elderly is mixing solely among themselves, and equal mixing between risk groups.
```
    config_params = {
        "r_0": 2.0,
        "infectious_period": 7.0,
        "age_demographics": jnp.array([0.7, 0.2, 0.1]),
        "risk_prop": jnp.array([[0.5, 0.5], [0.5, 0.5], [0.5, 0.5]]),
        "age_contact_matrix": jnp.array(
            [[0.8, 0.2, 0.0], [0.2, 0.8, 0.0], [0.0, 0.0, 1.0]]
        ),
        "risk_contact_matrix": jnp.array([[0.5, 0.5], [0.5, 0.5]]),
        "age_dimension": Dimension(
            name="age",
            bins=[
                AgeBin(0, 17, "young"),
                AgeBin(18, 64, "adult"),
                AgeBin(65, 99, "elderly"),
            ],
        ),
        "risk_dimension": Dimension(
            name="risk", bins=[Bin(name="high"), Bin(name="low")]
        ),
    }
    config = get_config(config_params=config_params)

    s0_prop = jnp.array([[0.99, 1.0], [0.99, 1.0], [1.0, 1.0]])
    i0_prop = jnp.array([[0.01, 0.0], [0.01, 0.0], [0.0, 0.0]])
```

Here's the run outcome (plotting only the susceptibles):
sir_age_risk_stratified
This seems to suggest that, despite equal mixing between risk groups, the infected in Young-High did not manage to infect Young-Low, and likewise for Adult-High -> Adult-Low. I suspect the FOI is not properly calculating this?

**Comment by kokbent** [comment]
I suspect the FOI calculation
https://github.com/CDCgov/DynODE/blob/976edceda3d2ec709fc294e5011b73179d45e6b9/examples/sir_age_risk_stratified.py#L153 
should be `"ijkl, ij -> kl"` instead, but I don't have any principled explanation of why except they seem to work in my dummy example here:

```
age_contact = np.array([[0.8, 0.2, 0.0], [0.2, 0.8, 0.0], [0.0, 0.0, 1.0]])
risk_contact = np.array([[0.6, 0.4], [0.4, 0.6]])

cm = np.einsum("ij, kl -> ikjl", age_contact, risk_contact)
i = np.array([[1.0, 0.0], [0.0, 0.0], [0.0, 0.0]])
np.einsum("ijkl, ij -> kl", cm, i)
# array([[0.48, 0.32],
#       [0.12, 0.08],
#       [0.  , 0.  ]])
```

You might want to check the calculation a lot more rigorously to be sure this is the right einsum to use.

**Comment by Fuhan-Yang** [comment]
Thanks Ben. The confusion here is that FOI for young and high-risk susceptible was calculated as $\sum_a\sum_r c_{ya}r_{hr}I_{yh}$ above (which has `np.einsum('ijkl, ij -> ij', (p.contact_matrix)/pop_size, i)`. This indicates that young and high-risk S only contacts with young and high-risk I. This depicts the strange dynamics you saw: each stratified population progresses independently as they only get infected by the population from the same stratification. So if the young and low-risk population has 0 infection at beginning, it won't have dynamics. 

The correct way is showed below, for young and high-risk S, their FOI should be the sum of their contacts with all other groups. This should be: $\sum_a\sum_r c_{ya}r_{hr}I_{ar}$, or $\sum_i\sum_j C_{ijkl}I_{ij}$ for all the sub-array in the 4-D contact matrix. And this gives us `np.einsum('ijkl, ij -> kl', (p.contact_matrix)/pop_size, i)`, exactly as you showed. This means summing over all the contacts across other groups for each age and risk combination.

image

Thank you Ben!

File: examples/sir_age_risk_stratified.py [hunk 7]
```diff
@@ -201,22 +215,23 @@ def sir_ode(
...
-            t,
-            i.reshape(-1, age_risk_dim)[:, idx],
-            label=f"Infectious ({label})",
-        )
-        plt.plot(
-            t,
-            r.reshape(-1, age_risk_dim)[:, idx],
-            label=f"Recovered ({label})",
-        )
+        # plt.plot(
```
Comment (kokbent) [inline examples/sir_age_risk_stratified.py]: Use matplotlib subplots to plot S, I, R in three panels? Could also just remove the comments otherwise

**Comment by kokbent** [comment]
Looks correct to me now...

(Although I struggle with the mathematical notions of FOI calculations: the one with a and r summations aren't really readable to me, the one with `ijkl` is actually a bit clearer... although having said that, those formulas would have yielded a scalar and not a matrix so it's not quite right. But it's not like we are writing the mathematical formulas into a document here so doesn't quite matter.)

**Comment by Fuhan-Yang** [comment]
The FOI will be a scalar from the last formula. After small tweaks, this script is able to generate the same dynamics as in `sir_age_stratified.py` given one risk bin and same parameters!
image

The script is then refactored to generate three subplots of S, I, R respectively. Let me know what else is needed to get this PR merged :)

**Comment by kokbent** [comment]
Need to update `CHANGELOG.md` and `pyproject.toml` or otherwise it would fail checks when merging. See [previous commit](https://github.com/CDCgov/DynODE/commit/b8d0939f2de355cf6c22d49772fc0e24b5bc9a4d) for the kind of changes expected. Can follow [this](https://github.com/CDCgov/DynODE/blob/main/CHANGELOG.md#202506162a---adding-more-examples-modules) for language on `CHANGELOG.md`.

**Review by kokbent — APPROVED** [review]
Looks good to go

**Comment by ElishaBayode** [comment]
Looks good to me too (for the records)!
